# Pandas for AI/ML – Conceptual & Technical Assessment



**Name:** Arege Sravya 

**Submission filename:** `sravya_Pandas.ipynb`


# Section A – Output Prediction

## Q1. Array operations and vectorization

In [24]:
import numpy as np
import pandas as pd

# Creating a NumPy array
arr = np.array([10, 20, 30])

# Creating a Pandas Series
s = pd.Series([10, 20, 30])

# NumPy performs element-wise multiplication
print(arr * 2)

# Pandas Series also performs element-wise multiplication
print(s * 2)

[20 40 60]
0    20
1    40
2    60
dtype: int64


### 1. Predict the output:

arr * 2 → [20 40 60]
s * 2 → Series containing 20, 40, 60 with index values 0, 1, 2.

### 2. Similarity:
Both NumPy arrays and Pandas Series support vectorized operations, so multiplication by 2 is performed on every element without using an explicit loop.

Difference:
A NumPy array does not normally have labeled indexes, whereas a Pandas Series has an index for each value.

## Q2. Boolean Filtering

In [25]:
import pandas as pd

# Creating a DataFrame containing student names and their marks
df = pd.DataFrame({
    'Student': ['A', 'B', 'C', 'D'],
    'Marks': [45, 80, 65, 90]
})

# Filtering the rows where Marks are greater than or equal to 65
result = df[df['Marks'] >= 65]

# Selecting only the Student and Marks columns
print(result[['Student', 'Marks']])

  Student  Marks
1       B     80
2       C     65
3       D     90


### 1. Predict the rows displayed:

The condition is:

df['Marks'] >= 65

It checks each student's marks:

A → 45 >= 65 → False
B → 80 >= 65 → True
C → 65 >= 65 → True
D → 90 >= 65 → True

Therefore, students B, C, and D are displayed.

### 2. Number of rows and columns:

Rows: 3
Columns: 2

So the resulting DataFrame has 3 rows and 2 columns.

Underlying Pandas operation

df['Marks'] >= 65 creates a Boolean Series:

0    False
1     True
2     True
3     True
Name: Marks, dtype: bool

When this Boolean Series is passed inside df[...], Pandas keeps only the rows where the condition is True. This is called Boolean filtering / Boolean indexing.

## Q3. Descriptive Statistics

In [26]:
import pandas as pd

# Create a DataFrame containing students' marks
df = pd.DataFrame({
    'Marks': [10, 20, 30, 40, 50]
})

# Calculate and print the average of all marks
print(df['Marks'].mean())

# Find and print the highest mark
print(df['Marks'].max())

# Find and print the middle value of the marks
print(df['Marks'].median())

30.0
50
30.0


### Explanation
mean() calculates the average of the values.
max() returns the highest value.
median() returns the middle value after sorting the values.

Mean vs Median: Mean is calculated using all values, while median is the middle value. The mean can be affected more by extreme values than the median.

## Q4. Group-wise Analysis

In [27]:
import pandas as pd

# Create a DataFrame containing departments and employee salaries
df = pd.DataFrame({
    'Department': ['IT', 'HR', 'IT', 'HR'],
    'Salary': [40000, 30000, 60000, 50000]
})

# Group employees by department and calculate the average salary
# for each department
print(df.groupby('Department')['Salary'].mean())

Department
HR    40000.0
IT    50000.0
Name: Salary, dtype: float64


groupby('Department') separates the employees into HR and IT groups. Then ['Salary'].mean() calculates the average salary separately for each department.

HR → (30000 + 50000) / 2 = 40000
IT → (40000 + 60000) / 2 = 50000

### Explanation: groupby() groups rows based on a column and applies an operation such as mean() to each group.

## Q5. Pandas Correlation – Positive

In [28]:
import pandas as pd

# Create a DataFrame containing study hours and marks
df = pd.DataFrame({
    'Hours': [1, 2, 3, 4],
    'Marks': [20, 40, 60, 80]
})

# Calculate the correlation between study hours and marks
print(df['Hours'].corr(df['Marks']))

1.0


### 1. Predict the correlation coefficient.

Answer: Correlation coefficient: 1.0

### 2. Interpret the direction and strength of the relationship

Answer: The correlation is positive and perfectly strong. As the number of study hours increases, the marks also increase in a perfectly linear patter

### Explanation: 
corr() calculates the Pearson correlation coefficient between the two Series, and 1.0 indicates a perfect positive linear relationship.

## Q6. Pandas Correlation – Negative

In [29]:
import pandas as pd

# Create a DataFrame containing product prices and demand
df = pd.DataFrame({
    'Price': [10, 20, 30, 40],
    'Demand': [100, 80, 60, 40]
})

# Calculate the correlation between price and demand
print(df['Price'].corr(df['Demand']))

-1.0


### 1. Predict the correlation coefficient
 Answer: -1.0

### 2. Explain what the sign of the coefficient indicates about the relations.

Answer: The correlation is negative and perfectly strong. As the price increases, the demand decreases in a perfectly linear pattern.

What does the sign indicate?

Positive (+) → both variables tend to increase together.
Negative (−) → when one variable increases, the other tends to decrease.
0 → no linear relationship.

Here, -1.0 indicates a perfect negative linear relationship between Price and Demand.

# Section B – Code Correction

### Q1. JSON and CSV

### 1. A CSV file contains a header row, but the following code incorrectly reads the first student's record as column names. Correct the code

In [30]:
import pandas as pd
from io import StringIO

# Example of the supplied CSV structure: the first row contains column names.
csv_data = '''Student,Marks
A,75
B,82
C,91'''

# Use the existing first row as the header instead of treating it as data.
df = pd.read_csv(StringIO(csv_data), header=0)

print(df)

  Student  Marks
0       A     75
1       B     82
2       C     91


**Issue:** `header=None` tells Pandas that there is no header.  
**Correction:** Use the first row as the header with `header=0` (or omit the argument).

### Q2. Data Analysis – Five Highest Salaries

In [31]:
import pandas as pd

df = pd.DataFrame({
    'Employee': ['A', 'B', 'C', 'D', 'E', 'F'],
    'Salary': [30000, 50000, 25000, 60000, 45000, 70000]
})

# Sort from highest salary to lowest and keep the first five records.
top_five = df.sort_values('Salary', ascending=False).head(5)

print(top_five)

  Employee  Salary
5        F   70000
3        D   60000
1        B   50000
4        E   45000
0        A   30000


**Issue:** Ascending sorting puts the lowest salaries first.  
**Correction:** Use `ascending=False` so the highest salaries come first, then select five rows.

### Q3. Cleaning Data – Missing Marks

In [32]:
import pandas as pd

df = pd.DataFrame({'Marks': [70, None, 90]})

# fillna() returns a filled Series, so assign it back to the DataFrame.
df['Marks'] = df['Marks'].fillna(df['Marks'].mean())

print(df)

   Marks
0   70.0
1   80.0
2   90.0


**Issue:** The original `fillna()` call did not store the returned result.  
**Correction:** Assign the filled values back to `df['Marks']`.

### Q4. Correlation

In [33]:
import pandas as pd

df = pd.DataFrame({
    'Hours': [1, 2, 3, 4],
    'Marks': [20, 40, 60, 80]
})

# Select the two Series and calculate their correlation.
correlation = df['Hours'].corr(df['Marks'])

print("Correlation:", correlation)

Correlation: 1.0


**Issue:** The original `DataFrame.corr()` call was given column names incorrectly.  
**Correction:** Select the two Series and call `Series.corr()`.

## Section C – Edge Cases

### Q1. Two NumPy arrays have shapes (3, 1) and (1, 3). Predict the shape of their element-wise sum and explain how broadcasting works. Why might the result
differ from adding two one-dimensional arrays?

In [34]:
import numpy as np

# Create a 2D column array with 3 rows and 1 column
a = np.array([[10], [20], [30]])

# Create a 2D row array with 1 row and 3 columns
b = np.array([[1, 2, 3]])

# NumPy broadcasts the two arrays to compatible shapes
# and performs element-wise addition
result = a + b

# Display the resulting array
print(result)

# Display the shape of the resulting array
print(result.shape)

[[11 12 13]
 [21 22 23]
 [31 32 33]]
(3, 3)


### 1. Predict the output
[[11 12 13]
 [21 22 23]
 [31 32 33]]
(3, 3)

### 2. Shapes before addition
a → (3, 1)
b → (1, 3)

### 3 Explain how NumPy broadcasting enables the addition of these arrays.

NumPy broadcasting automatically expands the smaller array to match the shape of the larger array, allowing element-wise addition without manually reshaping the arrays. Here, arrays of shapes (3,1) and (1,3) are broadcast to (3,3), and their corresponding elements are added.

### 4. What would happen if b were replaced with np.array([1, 2])? Explain the reason.

a → (3, 1)
b → (2,)

NumPy aligns dimensions from the right side:

a → (3, 1)
b →    (2)

The dimensions 1 and 2 are not compatible because neither is equal nor is either dimension 1 in the aligned comparison. Therefore, NumPy raises a ValueError instead of producing a result.

ValueError: operands could not be broadcast together with shapes (3,1) (2,)

### Explanation: 
(3,1) and (2,) cannot be broadcast together because their corresponding dimensions are incompatible.

### Q2. A CSV file contains commas inside quoted text fields, missing values, and dates in different formats. Explain how incorrect parsing can affect the
resulting DataFrame and identify suitable read options or preprocessing steps.



In [35]:
import numpy as np

a = np.array([[10], [20], [30]])
b = np.array([[1, 2, 3]])

# Broadcasting expands the compatible dimensions during element-wise addition.
result = a + b

print(result)
print("Shape:", result.shape)

# These shapes are incompatible for broadcasting.
try:
    a + np.array([1, 2])
except ValueError as error:
    print("Error:", error)

[[11 12 13]
 [21 22 23]
 [31 32 33]]
Shape: (3, 3)


### 1. How is "Asha, R" handled?

The comma inside "Asha, R" is correctly treated as part of the Name field because the value is enclosed in double quotes.

So Pandas reads it as:

ID   Name      Date        Marks
101  Asha, R   2026-09-01  80

It does not create an extra column for the comma.

### 2. Potential issues with Date and Marks columns
Date: The dates are in different formats: 2026-09-01 and 01/09/2026. They should be converted to a consistent datetime format.
Marks: The column contains missing values: one blank value and one "NA". These should be handled appropriately before numerical analysis.

### 3. Corrected code

In [36]:
import pandas as pd
from io import StringIO

csv_data = '''ID,Name,Date,Marks
101,"Asha, R",2026-09-01,80
102,Ravi,01/09/2026,
103,Neha,2026-09-03,NA'''

# Quoted commas stay inside the Name field.
# Blank values and NA are treated as missing values for inspection.
df = pd.read_csv(
    StringIO(csv_data),
    na_values=['NA', ''],
    keep_default_na=True
)

# Convert the mixed date representations into datetime values.
df['Date'] = pd.to_datetime(df['Date'], dayfirst=True, errors='coerce')

print(df)
print("\nData types:")
print(df.dtypes)

    ID     Name       Date  Marks
0  101  Asha, R 2026-01-09   80.0
1  102     Ravi        NaT    NaN
2  103     Neha 2026-03-09    NaN

Data types:
ID                int64
Name             object
Date     datetime64[ns]
Marks           float64
dtype: object


### 4. Why investigate "NA" before treating it as missing?

"NA" may represent a missing mark, but it could also be a meaningful value or a specific code used by the organization, such as Not Applicable or Not Assessed. Therefore, its meaning should be confirmed before replacing or removing it.

### Explanation: 
Always understand what a missing-value indicator means before cleaning it, because blindly replacing "NA" could lead to incorrect analysis.

### Q3. A salary column contains the values 25000, 30000, 35000, and 10000000. Explain how the extreme value affects the mean compared with the median.
Identify which statistic may better represent a typical employee's salary and justify your choice.

In [37]:
import pandas as pd

# Create a DataFrame containing employee salaries
df = pd.DataFrame({
    "Employee": ["A", "B", "C", "D"],
    "Salary": [25000, 30000, 35000, 10000000]
})

# Calculate and display the average salary
print("Mean:", df["Salary"].mean())

# Calculate and display the middle salary value
print("Median:", df["Salary"].median())

Mean: 2522500.0
Median: 32500.0


### 1. Predict the mean and median salary.

Mean: 2522500.0

Median: 32500.0

### 2. Identify the effect of the extreme salary on the mean.

The salary 10,000,000 is an extreme value compared with the other salaries. It pulls the mean significantly upward, making the mean much larger than the salaries of most employees.

The median is much less affected by this extreme value.

### 3. Which statistic may better represent a typical employee's salary in this dataset? Justify your answer.

The median may better represent the typical employee's salary in this dataset because it is less affected by the extremely high salary of 10,000,000.

### 4. Write code to identify the employee record with the highest salary and explain why the record should be investigated before removing it

In [38]:
import pandas as pd

# Create a DataFrame containing employee salaries
df = pd.DataFrame({
    "Employee": ["A", "B", "C", "D"],
    "Salary": [25000, 30000, 35000, 10000000]
})

# Find the index of the employee with the highest salary
highest_index = df["Salary"].idxmax()

# Display the complete record of the employee with the highest salary
print(df.loc[highest_index])

Employee           D
Salary      10000000
Name: 3, dtype: object


### Why investigate before removing it?
The value may be a genuine high salary, a data-entry error, or a special case such as a senior executive's compensation. It should be investigated and verified before deciding whether to remove it.

# Section D – Scenario-Based Problems

### 1. Load the CSV file and inspect the dataset using head(), info(), and isnull().sum().

In [1]:
import pandas as pd

# Load the student assessment dataset from the CSV file
df = pd.read_csv("student_assessment_dirty.csv")

# Display the first five rows of the dataset
print("First 5 rows:")
print(df.head())

# Display information about columns, data types, and missing values
print("\nDataset Information:")
df.info()

# Count the number of missing values in each column
print("\nMissing Values:")
print(df.isnull().sum())

First 5 rows:
  Student_ID Student_Name Department Assessment   Marks
0       S101         Asha         IT     Python      85
1       S102         Ravi         it     Python      78
2       S103         Neha       I.T.     Python      92
3       S104        Karan         HR     Python     NaN
4       S105        Meera         HR     Python  Absent

Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Student_ID    15 non-null     object
 1   Student_Name  15 non-null     object
 2   Department    15 non-null     object
 3   Assessment    15 non-null     object
 4   Marks         12 non-null     object
dtypes: object(5)
memory usage: 732.0+ bytes

Missing Values:
Student_ID      0
Student_Name    0
Department      0
Assessment      0
Marks           3
dtype: int64


### Explanation
head() displays the first five records.
info() shows column names, data types, and non-null values.
isnull().sum() counts missing values in each column.

### 2. Standardize department names so that IT, it, and I.T. are represented consistently. Apply the same approach to Finance.

In [2]:
# Remove extra spaces and convert all department names to uppercase
df["Department"] = df["Department"].str.strip().str.upper()

# Replace different forms of department names with standard names
df["Department"] = df["Department"].replace({
    "IT": "IT",
    "I.T.": "IT",
    "HR": "HR",
    "FINANCE": "Finance"
})

# Display the standardized department names
print(df["Department"].unique())

['IT' 'HR' 'Finance']


### Explanation

str.upper() makes different letter cases consistent, and replace() converts variations such as it, I.T., and finance into standard department names.

### 3. Identify and remove exact duplicate student assessment record

In [3]:
# Count the number of duplicate rows in the dataset
duplicate_count = df.duplicated().sum()

# Display the number of duplicate records
print("Number of duplicate records:", duplicate_count)

# Remove exact duplicate rows
df = df.drop_duplicates()

# Display the number of rows after removing duplicates
print("Number of records after removing duplicates:", len(df))

Number of duplicate records: 1
Number of records after removing duplicates: 14


### Explanation

duplicated() identifies rows that are completely identical, while drop_duplicates() removes the duplicate record.

### 4. Convert the Marks column into a numeric format, handling "88 marks", "Absent", "NA", and blank entries appropriately.

In [4]:
# Keep a copy of the original Marks column
# so that values such as "Absent" can be investigated later
df["Original_Marks"] = df["Marks"]

# Remove the word "marks" from values such as "88 marks"
df["Marks"] = (
    df["Marks"]
    .astype("string")
    .str.replace("marks", "", case=False, regex=False)
    .str.strip()
)

# Convert the cleaned Marks column into numeric values
# Non-numeric values such as "Absent" become NaN
df["Marks"] = pd.to_numeric(df["Marks"], errors="coerce")

# Display the Marks column after conversion
print(df[["Original_Marks", "Marks"]])

   Original_Marks  Marks
0              85     85
1              78     78
2              92     92
3             NaN   <NA>
4          Absent   <NA>
5               0      0
6              67     67
7              75     75
9             NaN   <NA>
10       88 marks     88
11             91     91
12            NaN   <NA>
13             64     64
14            105    105


### Explanation
"88 marks" becomes 88.
"Absent" becomes NaN.
"NA" is treated as missing.
Blank values remain missing.
Numeric values such as 85, 78, and 92 remain numbers.

### 5. Investigate missing marks and distinguish absent students from students whose marks have not been entered. State one risk of deleting missing values
without investigating why they are missing.

In [5]:
# Identify students who were explicitly marked as absent
absent_students = df[
    df["Original_Marks"].astype("string").str.strip().str.lower() == "absent"
]

# Display absent students
print("Absent Students:")
print(absent_students[["Student_ID", "Student_Name", "Original_Marks"]])

# Identify records where the cleaned Marks value is missing
missing_marks = df[df["Marks"].isna()]

# Display all records with missing marks
print("\nStudents with Missing Marks:")
print(missing_marks[["Student_ID", "Student_Name", "Original_Marks"]])

Absent Students:
  Student_ID Student_Name Original_Marks
4       S105        Meera         Absent

Students with Missing Marks:
   Student_ID Student_Name Original_Marks
3        S104        Karan            NaN
4        S105        Meera         Absent
9        S109        Sneha            NaN
12       S112         Aman            NaN


### Explanation

1.Absent should be retained as a separate status instead of treating it exactly like a mark that was never entered.

2.Risk of deleting missing values

3.Deleting missing values without investigating why they are missing may remove useful information and can lead to biased or incorrect analysis.

### 6. Identify marks outside the valid range of 0–100 and decide how they should be handled

In [7]:
# Find marks below 0 or above 100
invalid_marks = df[
    (df["Marks"] < 0) | (df["Marks"] > 100)
]

# Display the invalid records
print("Invalid Marks:")
print(invalid_marks[["Student_ID", "Student_Name", "Marks"]])

Invalid Marks:
   Student_ID Student_Name  Marks
14       S114        Kabir    105


### Handle the invalid mark

In [8]:
# Replace marks outside the valid range with a missing value
# until the original record is verified
df.loc[
    (df["Marks"] < 0) | (df["Marks"] > 100),
    "Marks"
] = pd.NA

### Explanation

105 should not automatically be deleted. It should first be investigated because it could be a data-entry error or a special case. Until it is verified, treating it as missing prevents it from affecting calculations.

### 7. Display the cleaned DataFrame and report the number of records removed, missing marks remaining, and invalid marks identified.

In [9]:
# Display the cleaned DataFrame
print("Cleaned DataFrame:")
print(df)

# Count the number of records removed
print("\nRecords removed:", duplicate_count)

# Count the number of missing marks remaining
print("Missing marks remaining:", df["Marks"].isna().sum())

# Count the number of invalid marks identified
print("Invalid marks identified:", len(invalid_marks))

Cleaned DataFrame:
   Student_ID Student_Name Department Assessment  Marks Original_Marks
0        S101         Asha         IT     Python     85             85
1        S102         Ravi         IT     Python     78             78
2        S103         Neha         IT     Python     92             92
3        S104        Karan         HR     Python   <NA>            NaN
4        S105        Meera         HR     Python   <NA>         Absent
5        S106        Arjun    Finance     Python      0              0
6        S107        Priya    Finance     Python     67             67
7        S108        Rahul         IT     Python     75             75
9        S109        Sneha         IT     Python   <NA>            NaN
10       S110        Varun         HR     Python     88       88 marks
11       S111         Diya    Finance     Python     91             91
12       S112         Aman         IT     Python   <NA>            NaN
13       S113        Nisha         HR     Python     64   

### Explanation

The dataset is cleaned by standardizing department names, removing exact duplicates, converting marks into numeric values, investigating missing/absent records, and identifying invalid marks outside the 0–100 range.